File: lib/game_machine/physics/grid.rb
module GameMachine
  module Physics

    # @note Thread safe 2d grid that tracks entity locations and  provides
    #   neighbor queries.  The grid stores 3d coordinates, but only calculates
    #   neighbors based on 2d.  
    # This class is thread safe.  You can create as many instances as you want.
    class Grid

      # max is the width of the grid, cell size should reflect the area
      #   that you want neighbor queries to use.
      def initialize(max,cell_size)
        @max = max
        @min = 0
        @cell_size = cell_size
        @conv_factor = 1.0/@cell_size
        @width = @max / @cell_size
        @cell_count = @width * @width
        @object_index = java.util.concurrent.ConcurrentHashMap.new
        @cells = java.util.concurrent.ConcurrentHashMap.new
        @cells_cache = java.util.concurrent.ConcurrentHashMap.new
      end

      # Returns the cells that are in range.
      def cells_within_radius(x,y,radius)
        key = "#{x}#{y}#{radius}"
        if cells = @cells_cache.get(key)
          return cells
        end

        offset = radius/@cell_size
        cells = [hash(x,y)]

        1.upto(offset) do |i|
          bounds = radius * i
          cells << hash(x - bounds,y - bounds)
          cells << hash(x - bounds,y + bounds)

          cells << hash(x + bounds,y - bounds)
          cells << hash(x + bounds,y + bounds)

          cells << hash(x - bounds,y)
          cells << hash(x + bounds,y)

          cells << hash(x,y - bounds)
          cells << hash(x,y + bounds)
        end

        cells.delete_if {|cell| cell < @min || cell > (@cell_count - 1)}
        result = cells.uniq
        unless @cells_cache.get(key)
          @cells_cache.put(key,result)
        end
        result
      end

      # Returns entities that are within range
      def neighbors(x,y,radius,type)
        result = {:players => [], :npcs => []}

        cells_within_radius(x,y,radius).each do |cell|
          points_in_cell(cell).each do |point|
            unless point.empty?
              point_class = point[2].class
              next if type != :all && type != point_class
              if point_class == Npc
                result[:npcs] << point[2]
              elsif point_class == Player
                result[:players] << point[2]
              end
            end
          end
        end
        result
      end

      def points_in_cell(cell)
        if points = @cells.get(cell)
          return points.values
        else
          return []
        end
      end

      def get(id)
        @object_index.get(id)
      end

      def remove(id)
        if values = @object_index.get(id)
          cell = values[1]
          if point = @cells.get(cell)
            point.remove(id)
          end
          @object_index.remove(id)
        end
      end

      # Set the location of an entity. Expects an entity with a transform
      #   and vector3 component.
      def set(entity)
        id = entity.id
        vector3 = entity.transform.vector3
        if existing_values = @object_index.fetch(id,nil)
          existing_vector3 = existing_values[2].transform.vector3
          if entity.transform.vector3.x == existing_vector3.x && entity.transform.vector3.y == existing_vector3.y
            return
          end
        end
        cell = hash(vector3.x,vector3.y)
        values = [id,cell,entity]
        @object_index[id] = values
        unless @cells.contains_key(cell)
          @cells.put(cell,java.util.concurrent.ConcurrentHashMap.new)
        end
        @cells.get(cell).put(id,values)
      end

      # Hashes an x,y pair to it's cell
      def hash(x,y)
        (x*@conv_factor).to_i + (y*@conv_factor).to_i * @width
      end

      def hash2(x,y)
        (x/@cell_size).floor + (y/@cell_size).floor * @width
      end

    end
  end
end

File: java/src/main/java/com/game_machine/core/Grid.java
package com.game_machine.core;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class Grid {

	private float max;
	private int cellSize = 0;
	private float convFactor;
	private int width;
	private int cellCount;
	private long lastNeighborCall;
	private long lastNeighborDelta;
	
	private ConcurrentHashMap<String, GridValue> objectIndex = new ConcurrentHashMap<String, GridValue>();
	private ConcurrentHashMap<Integer, ConcurrentHashMap<String, GridValue>> cells = new ConcurrentHashMap<Integer, ConcurrentHashMap<String, GridValue>>();
	private ConcurrentHashMap<Integer, Set<Integer>> cellsCache = new ConcurrentHashMap<Integer, Set<Integer>>();
	private ConcurrentHashMap<Integer, ArrayList<GridValue>> neighborsCache = new ConcurrentHashMap<Integer, ArrayList<GridValue>>();
	

	public Grid(int max, int cellSize) {
		this.max = max;
		this.cellSize = cellSize;
		this.convFactor = 1.0f / this.cellSize;
		this.width = (int) (this.max / this.cellSize);
		this.cellCount = this.width * this.width;
		lastNeighborCall = System.currentTimeMillis();
	}

	public Set<Integer> cellsWithinRadius(float x, float y) {
		int cellHash = hash(x, y);
		return cellsWithinRadius(cellHash,x,y);
	}
	
	public Set<Integer> cellsWithinRadius(int cellHash, float x, float y) {
		int key = cellHash;
		Set<Integer> cells = cellsCache.get(key);
		if (cells != null) {
			return cells;
		}
		cells = new HashSet<Integer>();

		int offset = this.cellSize;

		int startX = (int) (x - offset);
		int startY = (int) (y - offset);
		int endX = (int) (x + offset);
		int endY = (int) (y + offset);

		for (int rowNum = startX; rowNum <= endX; rowNum += this.cellSize) {
			for (int colNum = startY; colNum <= endY; colNum += this.cellSize) {
				if (rowNum >= 0 && colNum >= 0) {
					cells.add(hash(rowNum, colNum));
				}
			}
		}
		cellsCache.put(key, cells);
		return cells;
	}

	public ArrayList<GridValue> neighbors(float x, float y, String entityType) {
		int myCell = hash(x, y);
		return neighbors(myCell,x,y,entityType);
	}
	
	public ArrayList<GridValue> neighbors(int myCell, float x, float y, String entityType) {
		ArrayList<GridValue> result;

		lastNeighborDelta = System.currentTimeMillis() - lastNeighborCall;
		if (lastNeighborDelta <= 20) {
			result = neighborsCache.get(myCell);
			if (result != null) {
				return result;
			}
		}

		Collection<GridValue> gridValues;
		result = new ArrayList<GridValue>();
		Set<Integer> cells = cellsWithinRadius(myCell,x,y);
		for (int cell : cells) {
			gridValues = gridValuesInCell(cell);
			if (gridValues != null) {
				for (GridValue gridValue : gridValues) {
					if (entityType == null) {
						result.add(gridValue);
					} else if (gridValue.entityType.equals(entityType)) {
						result.add(gridValue);
					}
				}
			}
		}
		neighborsCache.put(myCell, result);
		lastNeighborCall = System.currentTimeMillis();
		return result;
	}

	public Collection<GridValue> gridValuesInCell(int cell) {
		ConcurrentHashMap<String, GridValue> cellGridValues = cells.get(cell);
		if (cellGridValues != null) {
			return cellGridValues.values();
		} else {
			return null;
		}
	}

	public GridValue get(String id) {
		return objectIndex.get(id);
	}

	public void remove(String id) {
		GridValue indexValue = objectIndex.get(id);
		if (indexValue != null) {
			int cell = indexValue.cell;
			ConcurrentHashMap<String, GridValue> cellGridValues = cells.get(cell);
			if (cellGridValues != null) {
				cellGridValues.remove(id);
			}
			objectIndex.remove(id);
		}
	}

	public Boolean set(String id, float x, float y, float z, String entityType) {
		GridValue oldValue = objectIndex.get(id);

		if (oldValue != null) {
			if (oldValue.x == x && oldValue.y == y) {
				return false;
			}
		}

		int cell = hash(x, y);
		if (!cells.containsKey(cell)) {
			cells.put(cell, new ConcurrentHashMap<String, GridValue>());
		}

		GridValue gridValue;
		gridValue = new GridValue(id, cell, x, y, z, entityType);

		if (oldValue != null && oldValue.cell != cell) {
			cells.get(oldValue.cell).remove(id);
		}
		cells.get(cell).put(id, gridValue);

		objectIndex.put(id, gridValue);

		return true;
	}

	public int hash2(float x, float y) {
		return (int) (Math.floor(x / this.cellSize) + Math.floor(y
				/ this.cellSize)
				* width);
	}

	public int hash(float x, float y) {
		return (int) ((x * this.convFactor)) + (int) ((y * this.convFactor))
				* this.width;
	}
}
